src/main/java/org/craftercms/engine/websocket/WsUpgradeHandler.java

```diff
@@ -0,0 +1,119 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.craftercms.engine.websocket;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpUpgradeHandler;
+import javax.servlet.http.WebConnection;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Socket;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+/**
+ * Upgrade websocket handler
+ */
+public class WsUpgradeHandler implements HttpUpgradeHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(WsUpgradeHandler.class);
+
+    ExecutorService exec;
+    InputStream socketIn;
+    OutputStream socketOut;
+    Socket socket;
+    Future<?> future;
+
+    public void preInit(ExecutorService exec, InputStream socketIn, OutputStream socketOut, Socket socket) {
+        this.exec = exec;
+        this.socketIn = socketIn;
+        this.socketOut = socketOut;
+        this.socket = socket;
+    }
+
+    /**
+     * Forward back and forth the websocket stream between websocket server and web client via engine as the proxy
+     * @param wc the WebConnection object associated to this upgrade request
+     */
+    @Override
+    public void init(WebConnection wc) {
+        logger.debug("* Websocket| Upgrade begin");
+        try {
+            var servletIn = wc.getInputStream();
+            var servletOut = wc.getOutputStream();
+            future = exec.submit(() -> {
+                logger.debug("> Websocket| Websocket server -> Engine");
+                int i = 0;
+                try {
+                    int bs;
+                    // Read data from Websocket server -> Engine and write to Client
+                    while ((bs = socketIn.read()) != -1) {
+                        servletOut.write(bs);
+                        servletOut.flush();
+                        i++;
+                    }
+                } catch (Exception exc) {
+                    logger.debug("> Websocket| Connection interrupted", exc);
+                } finally {
+                    servletOut.close();
+                }
+                logger.debug("< Websocket| Done: '{}'", i);
+                return null;
+            });
+
+            logger.debug("> Websocket| Client -> Engine");
+            int i = 0;
+            int bs;
+            try {
+                // Read data from Client -> Engine and write to Websocket server
+                while ((bs = servletIn.read()) != -1) {
+                    socketOut.write(bs);
+                    i++;
+                }
+            } catch (Exception exc) {
+                logger.debug("> Websocket| Connection interrupted", exc);
+            } finally {
+                socketOut.close();
+            }
+            logger.debug("> Websocket| Done: {}", i);
```

Comment:
This block seems the same as the one above in the task submitted to `future`. Can that be extracted to a method?
Also, are we just copying from an InputStream to an OutputStream? Consider something like [IOUtils.copy](https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/IOUtils.html#copy(java.io.InputStream,java.io.OutputStream)) or [InputStream.transferTo](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/InputStream.html#transferTo(java.io.OutputStream))

Reply:
@jmendeza I refactored this block to a new method `forwardStreamData`. Regarding the usage of `IOUtils.copy` and `InputStream.transferTo`, I have attempted to try these functions but did not successfully forward the socket. Some experiment shows that it is important to do the `outputStream.flush()` right after reading one block of data (255 bytes) to have the socket function normally. I have taken a look at these libraries' implementation and there is no `flush()` option after each block of reading so I kept the implementation of manually reading each data block of bytes.

Reply:
@jmendeza The `PrintStream` class has two `write` methods:

https://docs.oracle.com/javase/8/docs/api/java/io/PrintStream.html#write-int-

Both of them do not provide a way to flush after writing each byte to an output stream.
The one `write(int b)` seems to write a byte but only flush if the byte is a newline `\n` character. Here is the method implementation:
```
public void write(int b) {
        try {
            synchronized (this) {
                ensureOpen();
                out.write(b);
                if ((b == '\n') && autoFlush)
                    out.flush();
            }
        }
        catch (InterruptedIOException x) {
            Thread.currentThread().interrupt();
        }
        catch (IOException x) {
            trouble = true;
        }
    }
```

I also tried the following but was not successful forwarding the socket:
```
PrintStream printStream = new PrintStream(outputStream, flushOutput);
            while ((bs = inputStream.read()) != -1) {
                printStream.write(bs);
                i++;
            }
```